pkg/wmi/wmi.go

```diff
@@ -0,0 +1,130 @@
+// +build windows
+
+// Package wmi provides a basic interface for querying against
+// wmi. It's based on some underlying examples using ole [1].
+//
+// We do _not_ use the stackdriver library [2], because that uses reflect
+// and wants typed objects. Our use case is too dynamic.
+//
+// To understand the available classes, take a look at the Microsoft
+// documention [3]
+
+// References:
+//
+// 1. https://stackoverflow.com/questions/20365286/query-wmi-from-go
+// 2. https://github.com/StackExchange/wmi
+// 3. https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/operating-system-classes
+package wmi
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/go-ole/go-ole"
+	"github.com/go-ole/go-ole/oleutil"
+	"github.com/kolide/launcher/pkg/contexts/ctxlog"
+	"github.com/pkg/errors"
+)
+
+// S_FALSE is returned by CoInitializeEx if it was already called on this thread.
+const S_FALSE = 0x00000001
+
+func Query(ctx context.Context, className string, properties []string) ([]map[string]interface{}, error) {
+	logger := log.With(ctxlog.FromContext(ctx), "caller", "wmi.Query")
+	handler := NewOleHandler(ctx, properties)
+
+	// If we query for the exact fields, _and_ one of the property
+	// names is wrong, we get no results. (clearly an error. but I
+	// can't find it) So query for `*`, and then fetch the
+	// property. More testing might show this needs to change
+	queryString := fmt.Sprintf("SELECT * FROM %s", className)
```

Comment:
if the query specifies one of the properties that doesn't show up in `select *`, should we log that or throw an error? Perhaps that's better handled one level higher though..

Reply:
ah never mind, I see the handling for it below in HandleVariant

Reply:
It doesn't seem to matter. I'm not totally sure how it works under the WMI hood. But I think the higher level API is reasonable, and this may change